# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

In [3]:
target_text[0:400]

"new jersey est parfois calme pendant l' automne , et il est neigeux en avril .\nles états-unis est généralement froid en juillet , et il gèle habituellement en novembre .\ncalifornia est généralement calme en mars , et il est généralement chaud en juin .\nles états-unis est parfois légère en juin , et il fait froid en septembre .\nvotre moins aimé fruit est le raisin , mais mon moins aimé est la pomme"

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [4]:
def helper_func(text,dic,mode='source'):
    sentences = text.split('\n')
    ids = []
    for sentence in sentences:
        sentence_id = []
        if (mode=='target'):
            sentence = sentence + ' <EOS> '
        for word in sentence.split():
            sentence_id.append(dic[word])
        ids.append(sentence_id)
    return ids

def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_id_text = helper_func(source_text,source_vocab_to_int,mode='source')
    taget_id_text = helper_func(target_text,target_vocab_to_int,mode='target')
    return source_id_text, taget_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.4.1


/home/ahmed/Dropbox/udacity_dlfnp/venv/lib/python3.5/site-packages/ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [8]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(dtype=tf.int32,shape=[None,None],name='input')
    targets = tf.placeholder(dtype=tf.int32,shape=[None,None],name='target')
    lr = tf.placeholder(dtype=tf.float32,name='lr')
    keep_prob = tf.placeholder(dtype=tf.float32,name='keep_prob')
    target_sequence_length = tf.placeholder(dtype=tf.int32,shape=[None],name='target_sequence_length')
    max_target_len = tf.reduce_max(target_sequence_length,name='max_target_len')
    source_sequence_length = tf.placeholder(dtype=tf.int32,shape=[None],name='source_sequence_length')
    
    
    return inputs, targets, lr, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [9]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    ending = tf.strided_slice(input_=target_data, begin=[0, 0], end=[batch_size, -1], strides=[1, 1])
    decoder_input = tf.concat(values=[tf.fill(dims=[batch_size, 1], value=target_vocab_to_int['<GO>']), ending], axis=1)
    return decoder_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [10]:
from imp import reload
reload(tests)

def get_lstm_cell(rnn_size,keep_prob):
    lstm_cell = tf.contrib.rnn.BasicLSTMCell(num_units=rnn_size)
    lstm_cell_dropout = tf.contrib.rnn.DropoutWrapper(cell=lstm_cell,input_keep_prob=keep_prob)
    return lstm_cell
  
    
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    embed = tf.contrib.layers.embed_sequence(ids =rnn_inputs , vocab_size=source_vocab_size,embed_dim=encoding_embedding_size)
    
    cells = tf.contrib.rnn.MultiRNNCell(cells=[get_lstm_cell(rnn_size,keep_prob) for i in range(num_layers)])
    output, state = tf.nn.dynamic_rnn(cell=cells,inputs=embed,dtype=tf.float32)
    
    return output, state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [12]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """

    helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                               sequence_length=target_sequence_length,
                                              time_major=False)
    decoder = tf.contrib.seq2seq.BasicDecoder(cell=dec_cell,
                                              helper=helper,
                                              initial_state=encoder_state,
                                              output_layer=output_layer)
    final_outputs,_,_ = tf.contrib.seq2seq.dynamic_decode(decoder=decoder,
                                                            impute_finished=True,
                                                            maximum_iterations=max_summary_length)
    return final_outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [13]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    
    """

    start_tokens = tf.tile(input=tf.constant([start_of_sequence_id],dtype=tf.int32), 
                               multiples=[batch_size])

    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(embedding=dec_embeddings,
                                                     start_tokens=start_tokens,
                                                      end_token=end_of_sequence_id)
    
    decoder = tf.contrib.seq2seq.BasicDecoder(cell=dec_cell,
                                              helper=helper,
                                              initial_state=encoder_state,
                                              output_layer=output_layer)
    
    final_outputs,_,_= tf.contrib.seq2seq.dynamic_decode(decoder=decoder,
                                                    impute_finished=True,
                                                    maximum_iterations=max_target_sequence_length)

    return final_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [22]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """

    embedding = tf.Variable(initial_value=tf.random_uniform(shape=[target_vocab_size,decoding_embedding_size]))
    embed = tf.nn.embedding_lookup(params=embedding,ids=dec_input)
    
    
    dec_cell = tf.contrib.rnn.MultiRNNCell(cells=[get_lstm_cell(rnn_size,keep_prob) for i in range(num_layers)])
    output_layer = Dense(units=target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer())
    
    with tf.variable_scope("decode"):
        training_logits = decoding_layer_train(encoder_state, 
                                               dec_cell, 
                                               embed, 
                                               target_sequence_length, 
                                               max_target_sequence_length, 
                                               output_layer, 
                                               keep_prob)
    
    with tf.variable_scope("decode", reuse=True):
        inference_logits = decoding_layer_infer(encoder_state, 
                                                dec_cell, 
                                                embedding, 
                                                target_vocab_to_int['<GO>'],
                                                target_vocab_to_int['<EOS>'], 
                                                max_target_sequence_length, 
                                                target_vocab_size, 
                                                output_layer, 
                                                batch_size, 
                                                keep_prob)


            
        
    return training_logits,inference_logits
   
        

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [23]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    encoder_output, encoder_state = encoding_layer(input_data, 
                                   rnn_size, 
                                   num_layers, 
                                   keep_prob,
                                   source_sequence_length, 
                                   source_vocab_size, 
                                   enc_embedding_size)
    
    decoder_input = process_decoder_input(target_data, 
                          target_vocab_to_int, 
                          batch_size)
    
    training_logits,inference_logits = decoding_layer(decoder_input, 
                   encoder_state,
                   target_sequence_length, 
                   max_target_sentence_length,
                   rnn_size,
                   num_layers, 
                   target_vocab_to_int, 
                   target_vocab_size,
                   batch_size, 
                   keep_prob, 
                   dec_embedding_size)
    
    return training_logits,inference_logits
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [32]:
# Number of Epochs
epochs = 5
# Batch Size
batch_size = 50
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 100
decoding_embedding_size = 100
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 20

### Build the Graph
Build the graph using the neural network you implemented.

In [33]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

Batch and pad the source and target sequences

In [34]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [35]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})####


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   20/2757 - Train Accuracy: 0.3210, Validation Accuracy: 0.3818, Loss: 3.1504
Epoch   0 Batch   40/2757 - Train Accuracy: 0.3960, Validation Accuracy: 0.4545, Loss: 2.4085
Epoch   0 Batch   60/2757 - Train Accuracy: 0.4660, Validation Accuracy: 0.5064, Loss: 1.8295
Epoch   0 Batch   80/2757 - Train Accuracy: 0.4640, Validation Accuracy: 0.4873, Loss: 1.3734
Epoch   0 Batch  100/2757 - Train Accuracy: 0.4779, Validation Accuracy: 0.5509, Loss: 1.1766
Epoch   0 Batch  120/2757 - Train Accuracy: 0.5033, Validation Accuracy: 0.5245, Loss: 0.9997
Epoch   0 Batch  140/2757 - Train Accuracy: 0.5910, Validation Accuracy: 0.5936, Loss: 0.8970
Epoch   0 Batch  160/2757 - Train Accuracy: 0.5821, Validation Accuracy: 0.5718, Loss: 0.8065
Epoch   0 Batch  180/2757 - Train Accuracy: 0.6240, Validation Accuracy: 0.6009, Loss: 0.7241
Epoch   0 Batch  200/2757 - Train Accuracy: 0.5221, Validation Accuracy: 0.6009, Loss: 0.6984
Epoch   0 Batch  220/2757 - Train Accuracy: 0.5884, Validati

Epoch   0 Batch 1780/2757 - Train Accuracy: 0.9674, Validation Accuracy: 0.9473, Loss: 0.0587
Epoch   0 Batch 1800/2757 - Train Accuracy: 0.9362, Validation Accuracy: 0.9536, Loss: 0.0417
Epoch   0 Batch 1820/2757 - Train Accuracy: 0.9490, Validation Accuracy: 0.9518, Loss: 0.0300
Epoch   0 Batch 1840/2757 - Train Accuracy: 0.9705, Validation Accuracy: 0.9000, Loss: 0.0422
Epoch   0 Batch 1860/2757 - Train Accuracy: 0.9670, Validation Accuracy: 0.9491, Loss: 0.0293
Epoch   0 Batch 1880/2757 - Train Accuracy: 0.9610, Validation Accuracy: 0.9409, Loss: 0.0288
Epoch   0 Batch 1900/2757 - Train Accuracy: 0.9537, Validation Accuracy: 0.9509, Loss: 0.0400
Epoch   0 Batch 1920/2757 - Train Accuracy: 0.9189, Validation Accuracy: 0.9445, Loss: 0.0526
Epoch   0 Batch 1940/2757 - Train Accuracy: 0.9905, Validation Accuracy: 0.9591, Loss: 0.0253
Epoch   0 Batch 1960/2757 - Train Accuracy: 0.9370, Validation Accuracy: 0.9373, Loss: 0.0465
Epoch   0 Batch 1980/2757 - Train Accuracy: 0.9170, Validati

Epoch   1 Batch  800/2757 - Train Accuracy: 0.9568, Validation Accuracy: 0.9373, Loss: 0.0419
Epoch   1 Batch  820/2757 - Train Accuracy: 0.9780, Validation Accuracy: 0.9164, Loss: 0.0235
Epoch   1 Batch  840/2757 - Train Accuracy: 0.9210, Validation Accuracy: 0.9464, Loss: 0.0483
Epoch   1 Batch  860/2757 - Train Accuracy: 0.9232, Validation Accuracy: 0.8891, Loss: 0.0383
Epoch   1 Batch  880/2757 - Train Accuracy: 0.9580, Validation Accuracy: 0.9173, Loss: 0.0390
Epoch   1 Batch  900/2757 - Train Accuracy: 0.9240, Validation Accuracy: 0.9382, Loss: 0.0534
Epoch   1 Batch  920/2757 - Train Accuracy: 0.9470, Validation Accuracy: 0.9318, Loss: 0.0254
Epoch   1 Batch  940/2757 - Train Accuracy: 0.9540, Validation Accuracy: 0.9636, Loss: 0.0294
Epoch   1 Batch  960/2757 - Train Accuracy: 0.9830, Validation Accuracy: 0.9782, Loss: 0.0504
Epoch   1 Batch  980/2757 - Train Accuracy: 0.9467, Validation Accuracy: 0.9082, Loss: 0.0679
Epoch   1 Batch 1000/2757 - Train Accuracy: 0.9210, Validati

Epoch   1 Batch 2560/2757 - Train Accuracy: 0.9533, Validation Accuracy: 0.9182, Loss: 0.0583
Epoch   1 Batch 2580/2757 - Train Accuracy: 0.9781, Validation Accuracy: 0.9345, Loss: 0.0374
Epoch   1 Batch 2600/2757 - Train Accuracy: 0.9580, Validation Accuracy: 0.9055, Loss: 0.0308
Epoch   1 Batch 2620/2757 - Train Accuracy: 0.9411, Validation Accuracy: 0.9418, Loss: 0.0316
Epoch   1 Batch 2640/2757 - Train Accuracy: 0.9621, Validation Accuracy: 0.9355, Loss: 0.0236
Epoch   1 Batch 2660/2757 - Train Accuracy: 0.9895, Validation Accuracy: 0.9464, Loss: 0.0269
Epoch   1 Batch 2680/2757 - Train Accuracy: 0.9230, Validation Accuracy: 0.9564, Loss: 0.0310
Epoch   1 Batch 2700/2757 - Train Accuracy: 0.9720, Validation Accuracy: 0.9682, Loss: 0.0317
Epoch   1 Batch 2720/2757 - Train Accuracy: 0.9621, Validation Accuracy: 0.9409, Loss: 0.0296
Epoch   1 Batch 2740/2757 - Train Accuracy: 0.9905, Validation Accuracy: 0.9482, Loss: 0.0223
Epoch   2 Batch   20/2757 - Train Accuracy: 0.9700, Validati

Epoch   2 Batch 1580/2757 - Train Accuracy: 0.9947, Validation Accuracy: 0.9600, Loss: 0.0276
Epoch   2 Batch 1600/2757 - Train Accuracy: 0.9670, Validation Accuracy: 0.9191, Loss: 0.0198
Epoch   2 Batch 1620/2757 - Train Accuracy: 0.9480, Validation Accuracy: 0.9636, Loss: 0.0357
Epoch   2 Batch 1640/2757 - Train Accuracy: 0.9620, Validation Accuracy: 0.9573, Loss: 0.0167
Epoch   2 Batch 1660/2757 - Train Accuracy: 0.9779, Validation Accuracy: 0.9600, Loss: 0.0297
Epoch   2 Batch 1680/2757 - Train Accuracy: 0.9440, Validation Accuracy: 0.9382, Loss: 0.0449
Epoch   2 Batch 1700/2757 - Train Accuracy: 0.9600, Validation Accuracy: 0.9418, Loss: 0.0297
Epoch   2 Batch 1720/2757 - Train Accuracy: 0.9711, Validation Accuracy: 0.9555, Loss: 0.0314
Epoch   2 Batch 1740/2757 - Train Accuracy: 0.9663, Validation Accuracy: 0.9764, Loss: 0.0405
Epoch   2 Batch 1760/2757 - Train Accuracy: 0.9950, Validation Accuracy: 0.9536, Loss: 0.0159
Epoch   2 Batch 1780/2757 - Train Accuracy: 0.9432, Validati

Epoch   3 Batch  600/2757 - Train Accuracy: 0.9470, Validation Accuracy: 0.9382, Loss: 0.0427
Epoch   3 Batch  620/2757 - Train Accuracy: 0.9800, Validation Accuracy: 0.9445, Loss: 0.0271
Epoch   3 Batch  640/2757 - Train Accuracy: 0.9700, Validation Accuracy: 0.9309, Loss: 0.0285
Epoch   3 Batch  660/2757 - Train Accuracy: 0.9533, Validation Accuracy: 0.9600, Loss: 0.0251
Epoch   3 Batch  680/2757 - Train Accuracy: 0.9290, Validation Accuracy: 0.9727, Loss: 0.0235
Epoch   3 Batch  700/2757 - Train Accuracy: 0.9305, Validation Accuracy: 0.9636, Loss: 0.0426
Epoch   3 Batch  720/2757 - Train Accuracy: 0.9250, Validation Accuracy: 0.9309, Loss: 0.0272
Epoch   3 Batch  740/2757 - Train Accuracy: 0.9853, Validation Accuracy: 0.9318, Loss: 0.0130
Epoch   3 Batch  760/2757 - Train Accuracy: 0.9476, Validation Accuracy: 0.9355, Loss: 0.0256
Epoch   3 Batch  780/2757 - Train Accuracy: 0.9516, Validation Accuracy: 0.9645, Loss: 0.0534
Epoch   3 Batch  800/2757 - Train Accuracy: 0.9684, Validati

Epoch   3 Batch 2360/2757 - Train Accuracy: 0.9547, Validation Accuracy: 0.9218, Loss: 0.0559
Epoch   3 Batch 2380/2757 - Train Accuracy: 0.9467, Validation Accuracy: 0.9382, Loss: 0.0416
Epoch   3 Batch 2400/2757 - Train Accuracy: 0.9326, Validation Accuracy: 0.9273, Loss: 0.0744
Epoch   3 Batch 2420/2757 - Train Accuracy: 0.9578, Validation Accuracy: 0.9545, Loss: 0.0206
Epoch   3 Batch 2440/2757 - Train Accuracy: 0.9568, Validation Accuracy: 0.9600, Loss: 0.0212
Epoch   3 Batch 2460/2757 - Train Accuracy: 0.9514, Validation Accuracy: 0.9564, Loss: 0.0217
Epoch   3 Batch 2480/2757 - Train Accuracy: 0.9295, Validation Accuracy: 0.9473, Loss: 0.0416
Epoch   3 Batch 2500/2757 - Train Accuracy: 0.9979, Validation Accuracy: 0.9345, Loss: 0.0243
Epoch   3 Batch 2520/2757 - Train Accuracy: 0.9690, Validation Accuracy: 0.9491, Loss: 0.0180
Epoch   3 Batch 2540/2757 - Train Accuracy: 0.9680, Validation Accuracy: 0.9355, Loss: 0.0236
Epoch   3 Batch 2560/2757 - Train Accuracy: 0.9756, Validati

Epoch   4 Batch 1380/2757 - Train Accuracy: 0.9780, Validation Accuracy: 0.9573, Loss: 0.0212
Epoch   4 Batch 1400/2757 - Train Accuracy: 0.9840, Validation Accuracy: 0.9564, Loss: 0.0246
Epoch   4 Batch 1420/2757 - Train Accuracy: 0.9053, Validation Accuracy: 0.9491, Loss: 0.0399
Epoch   4 Batch 1440/2757 - Train Accuracy: 0.9630, Validation Accuracy: 0.9582, Loss: 0.0164
Epoch   4 Batch 1460/2757 - Train Accuracy: 0.9550, Validation Accuracy: 0.9482, Loss: 0.0182
Epoch   4 Batch 1480/2757 - Train Accuracy: 0.9516, Validation Accuracy: 0.9673, Loss: 0.0424
Epoch   4 Batch 1500/2757 - Train Accuracy: 0.9610, Validation Accuracy: 0.9464, Loss: 0.0262
Epoch   4 Batch 1520/2757 - Train Accuracy: 0.9667, Validation Accuracy: 0.9464, Loss: 0.0381
Epoch   4 Batch 1540/2757 - Train Accuracy: 0.9838, Validation Accuracy: 0.9336, Loss: 0.0219
Epoch   4 Batch 1560/2757 - Train Accuracy: 0.9695, Validation Accuracy: 0.9700, Loss: 0.0185
Epoch   4 Batch 1580/2757 - Train Accuracy: 0.9726, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [54]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    sentence = sentence.lower()
    out = []
    for word in sentence.split():
        if word in vocab_to_int:
            out.append(vocab_to_int[word])
        else:
            out.append(vocab_to_int['<UNK>'])

    return out


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [59]:
# translate_sentence = 'he saw a old yellow truck .'
translate_sentence = 'he is driving the red car .'

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [167, 143, 144, 228, 157, 207, 123]
  English Words: ['he', 'is', 'driving', 'the', 'red', 'car', '.']

Prediction
  Word Ids:      [250, 270, 21, 78, 50, 219, 1]
  French Words: il conduit le camion rouge . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.